core/src/main/java/com/linecorp/armeria/common/CookieUtil.java

```diff
@@ -0,0 +1,281 @@
+                throw new IllegalArgumentException("Cookie name contains an invalid char: " + name.charAt(pos));
+            }
+
+            final CharSequence unwrappedValue = unwrapValue(value);
+            if (unwrappedValue == null) {
+                throw new IllegalArgumentException("Cookie value wrapping quotes are not balanced: " + value);
+            }
+
+            if ((pos = firstInvalidCookieValueOctet(unwrappedValue)) >= 0) {
+                throw new IllegalArgumentException("Cookie value contains an invalid char: " +
+                                                   unwrappedValue.charAt(pos));
+            }
+        }
+    }
+
+    // Forked from netty-4.1.43
+    // https://github.com/netty/netty/blob/97d871a7553a01384b43df855dccdda5205ae77a/codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java
+
+    @Nullable
+    static CookieBuilder initCookie(Logger logger, boolean strict,
+                                    String header, int nameBegin, int nameEnd, int valueBegin, int valueEnd) {
+        if (nameBegin == -1 || nameBegin == nameEnd) {
+            logger.debug("Skipping cookie with null name");
+            return null;
+        }
+
+        if (valueBegin == -1) {
+            logger.debug("Skipping cookie with null value");
+            return null;
+        }
+
+        final CharSequence wrappedValue = CharBuffer.wrap(header, valueBegin, valueEnd);
+        final CharSequence unwrappedValue = unwrapValue(wrappedValue);
+        if (unwrappedValue == null) {
+            logger.debug("Skipping cookie because starting quotes are not properly balanced in '{}'",
+                         wrappedValue);
+            return null;
+        }
+
+        final String name = header.substring(nameBegin, nameEnd);
+
+        int invalidOctetPos;
+        if (strict && (invalidOctetPos = firstInvalidCookieNameOctet(name)) >= 0) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Skipping cookie because name '{}' contains invalid char '{}'",
+                             name, name.charAt(invalidOctetPos));
+            }
+            return null;
+        }
+
+        final boolean valueQuoted = unwrappedValue.length() != valueEnd - valueBegin;
+
+        if (strict && (invalidOctetPos = firstInvalidCookieValueOctet(unwrappedValue)) >= 0) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Skipping cookie because value '{}' contains invalid char '{}'",
+                             unwrappedValue, unwrappedValue.charAt(invalidOctetPos));
+            }
+            return null;
+        }
+
+        return Cookie.builder(name, unwrappedValue.toString()).valueQuoted(valueQuoted);
+    }
+
+    // The methods newly added in the fork.
+
+    static Cookies fromSetCookieHeaders(ImmutableSet.Builder<Cookie> builder,
+                                        boolean strict, Iterator<String> it) {
+        do {
```

Comment:
I think this do / while is to save a single call to `hasNext`, whereas usually we're not sure if the `it` is empty or not and use a normal while. I'm not sure this optimization is enough to not use a normal iteration style though. If we really wanted to keep it, I'd probably still add `assert it.hasNext()` at the beginning to make clear why we're using do/while here.

Reply:
Good idea. Added an assertion.